core/pta/widevine.c

```diff
@@ -0,0 +1,199 @@
+	}
+
+#define TPM_AUTH_PUB_MAX_SIZE 512
+#define WIDEVINE_PRIV_MAX_SIZE 32
+
+#define CROS_HWSEC_TA_UUID                                             \
+	{                                                              \
+		0xed800e33, 0x3c58, 0x4cae,                            \
+		{                                                      \
+			0xa7, 0xc0, 0xfd, 0x16, 0x0e, 0x35, 0xe0, 0x0d \
+		}                                                      \
+	}
+#define CROS_HDCP_PROV4_TA_UUID                                        \
+	{                                                              \
+		0x0feb839c, 0xee25, 0x4920,                            \
+		{                                                      \
+			0x8e, 0xe3, 0xac, 0x8d, 0xaa, 0x86, 0x0d, 0x3b \
+		}                                                      \
+	}
+#define TA_OPTEE_OEMCRYPTO_UUID                                        \
+	{                                                              \
+		0xa92d116c, 0xce27, 0x4917,                            \
+		{                                                      \
+			0xb3, 0x0c, 0x4a, 0x41, 0x6e, 0x2d, 0x93, 0x51 \
+		}                                                      \
+	}
+
+static const TEE_UUID allowed_ta_uuids[3] = {
+	CROS_HWSEC_TA_UUID,
+	CROS_HDCP_PROV4_TA_UUID,
+	TA_OPTEE_OEMCRYPTO_UUID,
+};
+
+/*
+ * The TPM auth public key. Used to communicate the TPM from OP-TEE.
+ * The format of data should be TPM2B_PUBLIC.
+ * For more information, please reference the 12.2.5 section:
+ * https://trustedcomputinggroup.org/wp-content/uploads/TCG_TPM2_r1p59_Part2_Structures_pub.pdf
+ */
+static uint8_t tpm_auth_pub[TPM_AUTH_PUB_MAX_SIZE];
+static uint32_t tpm_auth_pub_size;
+
+/*
+ * The Widevine root of trust secret. Used to sign the widevine
+ * request in OP-TEE. The value is an ECC NIST P-256 scalar.
+ * For more information, please reference the G.1.2 section:
+ * https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-186.pdf
+ */
+static uint8_t widevine_priv[WIDEVINE_PRIV_MAX_SIZE];
+static uint32_t widevine_priv_size;
+
+static TEE_Result init_widevine_dt_data(void)
+{
+	int node = 0;
+	int len = 0;
+	void *fdt = NULL;
+	const void *value = NULL;
+
+	fdt = get_dt();
+	if (!fdt)
+		return TEE_ERROR_NO_DATA;
+
+	node = fdt_path_offset(fdt, "/options/widevine");
+	if (node < 0)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	value = fdt_getprop(fdt, node, "tcg-tpm-auth-public-key", &len);
+	if (!value)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	if (len > TPM_AUTH_PUB_MAX_SIZE)
+		return TEE_ERROR_OVERFLOW;
+
+	memcpy(tpm_auth_pub, value, len);
+	tpm_auth_pub_size = len;
+
+	value = fdt_getprop(fdt, node, "widevine-root-of-trust", &len);
+	if (!value)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	if (len > WIDEVINE_PRIV_MAX_SIZE)
+		return TEE_ERROR_OVERFLOW;
+
+	memcpy(widevine_priv, value, len);
+	widevine_priv_size = len;
+
+	return TEE_SUCCESS;
+}
+
+service_init(init_widevine_dt_data);
+
+static TEE_Result verify_caller(void)
```

Comment:
I think this should be done from session opening. REE client should not be allowed to open a session towards this PTA.